src/main/scala/chisel3/tester/legacy/backends/vcs/IntervalTest.scala

```diff
@@ -0,0 +1,96 @@
+// See README.md for license details.
+
+package chisel3.tests
+
+// See README.md for license details.
+
+import chisel3._
+import chisel3.experimental._
+import chisel3.internal.firrtl.IntervalRange
+import chisel3.tester._
+import chisel3.tester.experimental.TestOptionBuilder._
+import chisel3.tester.internal.TreadleBackendAnnotation
+import logger.LazyLogging
+import org.scalatest._
+
+/** Shows several usages of Interval working converting larger ranges to smaller ones
+  * To see tabular results useful for debugging add the following line
+  *   {{{
+  *       _root_.logger.Logger.setLevel("chisel3.tests.IntervalTest", _root_.logger.LogLevel.Debug)
+  *   }}}
+  * to the line after {{{class IntervalTest}}}
+  */
+
+class HasIntervalWithBinaryPoint(inputRange: IntervalRange, outputRange: IntervalRange) extends Module {
+  val io = IO(new Bundle {
+    val input = Input(Interval(inputRange))
+
+    val out1 = Output(Interval(inputRange))
+    val out1Clip = Output(Interval(outputRange))
+    val out1Squeeze = Output(Interval(outputRange))
+    val out1Wrap = Output(Interval(outputRange))
+  })
+
+  io.out1 := io.input
+  io.out1Clip := io.input.clip(io.out1Clip)
+  io.out1Squeeze := io.input.squeeze(io.out1Squeeze)
+  io.out1Wrap := io.input.wrap(io.out1Wrap)
+}
+
+class IntervalTest extends FreeSpec with ChiselScalatestTester with LazyLogging with Matchers {
+
+  def intervalTestImpl(inputRange: IntervalRange, outputRange: IntervalRange): Unit = {
+
+    def wrap(b: BigDecimal): BigDecimal = {
+      val minValue: BigDecimal = outputRange.getLowestPossibleValue.get
+      val maxValue: BigDecimal = outputRange.getHighestPossibleValue.get
+      val delta:    BigDecimal = inputRange.increment.get
+      val lowToHigh = (maxValue - minValue) + delta
+      if (b < minValue) {
+        lowToHigh + b
+      } else if (b > maxValue) {
+        -lowToHigh + b
+      } else {
+        b
+      }
+    }
+
+    val inputBinaryPoint = inputRange.binaryPoint.get
+    val outputBinaryPoint = outputRange.binaryPoint.get
+
+    test(new HasIntervalWithBinaryPoint(inputRange, outputRange)).withAnnotations(Seq(TreadleBackendAnnotation)) { c =>
+      logger.debug(f"${"in"}%10s   " + f"${"Clip"}%10s   " + f"${"Squeeze"}%10s   " + f"${"Wrap"}%10s")
+
+      for (i <- inputRange.getPossibleValues) {
```

Comment:
Not a huge fan of tests where the expected value is a chain of complex logic from the inputs, because there's a higher chance of the test being wrong from bugs in the logic. I also don't think the generator nature in this specific case buys you much: you aren't exhaustively testing anything, so it might make more sense to write a few testvectors that address the common cases (zero, one, two, another power of two, another non-power-of-two).

Furthermore, this doesn't need to test Intervals as a feature, only that the testbench can interface with them. I think I would have written something similar to the FixedPoint test in [ElementTest.scala](https://github.com/ucb-bar/chisel-testers2/blob/master/src/test/scala/chisel3/tests/ElementTest.scala)

Reply:
Each of these tests turned up unexpected bugs in both the Interval implementation and in the tools that let one iterate over the values which. If I could reasonably anticipate what the edge cases are I could write a bunch of individual test, but I don't think that's easy to do. Having software models for things can be a problem but it's reasonable approach to testing that your code does what you think it does. Firrtl fixes might reveal errors in the model but I'd like to know when the model is flawed.

Reply:
So does this test belong here, or in chisel3 core, or in firrtl, or in treadle? The tests here aren't meant to be the chisel3 regression suite - and having such tests here means they aren't run as part of the chisel3 pre-merge test suite.

It might be useful in that it is testing something, but I don't think most of what this is testing is in this repo?